### PR TITLE
infosync: try send to leader after HTTP response not 200

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #53528

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/domain/infosync/info.go [hunk 1]
```diff
@@ -501,13 +510,57 @@ func doRequest(ctx context.Context, apiName string, addrs []string, route, metho
...
+				var (
+					leaderAddr string
+				)
+				for _, addr2 := range addrs {
+					status, err2 := etcdCli.Status(ctx, addr2)
+					if err2 != nil {
+						logutil.BgLogger().Warn(
+							"failed to get leader by checking endpoint status",
+							zap.String("endpoint", addr2),
+							zap.Error(err2))
```
Comment (JmPotato) [inline pkg/domain/infosync/info.go:522]: Will this log become overwhelming if the network fails?

File: pkg/domain/infosync/info.go [hunk 2]
```diff
@@ -501,13 +510,57 @@ func doRequest(ctx context.Context, apiName string, addrs []string, route, metho
...
+
+				logutil.BgLogger().Warn("response not 200, will retry sending to etcd leader",
 					zap.String("method", method),
 					zap.String("hosts", addr),
 					zap.String("url", url),
 					zap.Int("http status", res.StatusCode),
 					zap.Int("address order", idx),
+					zap.String("leader", leaderAddr),
 				)
+				if leaderAddr != "" {
```
Comment (JmPotato) [inline pkg/domain/infosync/info.go:539]: If `leaderAddr` is the same as the requested address, should we still make the request?

Comment (lance6716) [reply pkg/domain/infosync/info.go]: https://github.com/pingcap/tidb/pull/56332/commits/1e74686074b7bb90017bed7f4c93cec43ffadb98

File: pkg/domain/infosync/info.go [hunk 3]
```diff
@@ -501,13 +511,67 @@ func doRequest(ctx context.Context, apiName string, addrs []string, route, metho
...
+					}
+					if body != nil {
+						req2.Header.Set("Content-Type", "application/json")
+					}
+					res2, err2 := doRequestWithFailpoint(req2)
+					if err2 == nil {
+						bodyBytes2, err2 := io.ReadAll(res2.Body)
+						terror.Log(res2.Body.Close())
+						if err2 != nil {
+							return nil, err2
```
Comment (D3Hunter) [inline pkg/domain/infosync/info.go:563]: should we keep try other addresses here like previous err2?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/domain/infosync/info_test.go [hunk 4]
```diff
@@ -319,3 +319,43 @@ func TestRuleOp(t *testing.T) {
...
+			require.NoError(t, err)
+			if status.Leader == status.Header.MemberId {
+				leader = addr[len("unix://"):]
+				return true
+			}
+		}
+		return false
+	}, 5*time.Second, 100*time.Millisecond)
+
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/domain/infosync/FailPlacement", `1*return(true)`))
```
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/domain/infosync/info_test.go:350]: release-7.5 does not have this package

File: pkg/domain/infosync/info.go (hunk 3 above)
Comment (lance6716) [reply pkg/domain/infosync/info.go:563]: By the design we only try to connect PD leader (the same as etcd leader) to solve the oncall issue.

**Review by D3Hunter — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Pull request merged**